use dhcpcd instead of udhcpc

abom opened the pull request:
Related:

#1793


Testing replacing older running `networkd`:

```
~ # zinit status dhcp-npub4
---
name: dhcp-npub4
pid: 967
state: Running
target: Up
after: {}
~ # cat /etc/zinit/dhcp-npub4.yaml 
exec: ip netns exec ndmz /sbin/udhcpc -v -f -i npub4 -t 20 -T 1 -s /usr/share/udhcp/simple.script
~ # zinit status dhcp-npub4
---
name: dhcp-npub4
pid: 1282
state: Running
target: Up
after: {}
~ # cat /etc/zinit/dhcp-npub4.yaml 
exec: ip netns exec ndmz /usr/sbin/dhcpcd npub4 -B
```

abom committed:
use dhcpcd instead of udhcpc

abom committed:
keep udhcp for probing

abom committed:
terminate older udhcpc service if found

muhamadazmy commented on cmds/internet/main.go:
```diff
@@ -124,10 +126,16 @@ func configureZOS() error {
+		if err := z.Terminate(dhcpService, "udhcpc"); err != nil {
+			log.Error().Err(err).Msg("fail to terminate older dhcp-zos (udhcpc) zinit service")
+			return err
+		}
```
Terminate is not enough. since zinit still know the service was that specific name, replacing the zinit yaml file will not override the service configuration in zinit memory.
in other words, when u terminate a service, then change the file then start it will still start with the original config (before replacing the file)
Instead, what you need to do is to `terminate`, `forget`, `change the config file`, then `monitor`.

That being said, the init service does not live forever, it starts one time on system boot and is included in the basic zos image. It means that internet service does NOT need to check or replace the config, because there is nothing before it starts, which means it can directly use the dhcpcd service. 

It's up to `networkd` which can be updated during the machine runtime to check if the dhcp service is in fact using the dhcpcd binary, otherwise it then should terminate, forget, repalce, and monitor using the new config.

muhamadazmy replied:
I just noticed you added the new `Terminate` function to which does exactly that. which is nice, but i would rename it to something like `Destroy` instead of terminate.

I still think `internet` does not need to care since it's one time thing that starts on boot, hence there is no service before already.

muhamadazmy commented on pkg/zinit/commands.go:
```diff
@@ -304,3 +304,27 @@ func (c *Client) Forget(service string) error {
 }
+
+// Terminate a service completely (stop, forget and remove) if it matches an exec pattern
+func (c *Client) Terminate(service string, execPattern string) error {
```
I have some concerns about this method. It's very use case specific that it shouldn't be part of the client. I mean there is no actual use case for it outside this PR. Hence i suggest that we need to either generalize the use case, or move it outside of the client.

Of course generalizing the use case is the best thing to do. And i suggest the following "interface"

```go
func (c *Client) Destroy(filter ...Filter) error {
   // get all services, and for each service that matches ALL the filters, 
   // apply the destroy. operation
}

type Filter interface{
    matches(service *Service) bool
}

type nameFilter struct{
  name string
  }
// implement FIlter for the nameFilter type
// add other filter like `exe` filter  

// then u can call the Destroy like this
client.Destoy(WithName("dhcp-zos"), WithExe("udhcpc"))

// also if filters list is empty the method should return an error that search pattern is too wide
```

muhamadazmy commented on pkg/zinit/commands.go:
```diff
@@ -304,3 +304,27 @@ func (c *Client) Forget(service string) error {
+		return nil
+	}
+
+	if !strings.Contains(initService.Exec, execPattern) {
```
contains is not a good way to match on an exec instead do `filepath.Base(initService.Exec) == exec` to match exactly on the binary name. If you wanna implement a very powerful pattern matching then use regex. the problem with contains is that if your pattern is something like "bin" then you will kill all services.

so we can have 2 filters implemented
1) WithExec("bin-name") need to be documented this match on the executable name (not the full path)
2) WithExecRegex("<regex-pattern>") full regex match on the entire exec (not needed to be implemented now but can be easily added in the future.

Also since u pass the entire service object to the filter, it's easy now to build filters that also matches on other services patterns.

Even a better design for this api is to have 2 functions 
```
func (c *Client) Matches(filter ...Filter) ([]Service, error) 
// and 
func (c *Client) Destroy(wait, ...string) error
```

Which means filtering on service and destroying are 2 separate operations. So you can find services to do something else with and not always destroy

abom replied:
IMHO think that `Contains` is enough and it ony matches the exec of a single service and we won't allow a match with `bin` for example to be merged I think :laughing:.

But, I started to implement your suggestion, but got some concerns

* `Exec` field of `InitService` does not actually contain a "path" only (most of the time), it might even contain a shell script.
* `InitService` does not have a name field, but I think adding it would be easy (or using service name directly) and won't affect anything else.
* As contains, how would we control the input of `Regex` matches?

So, I suggest we can add these filtering check, but only apply it to exact "service" with a given name, not all services.

abom committed:
wip: update method name and implement filters

abom committed:
fix matches() call

muhamadazmy commented on pkg/zinit/commands.go:
```diff
@@ -304,3 +323,29 @@ func (c *Client) Forget(service string) error {
 }
+
+// Destroy a service completely (stop, forget and remove) if all filters are matched
+func (c *Client) Destroy(service string, filters ...Filter) error {
```
I suggested also that the searching `the matching` and `the destroying` are to be done sepeartely. So you have a Match method, and the Destroy method that takes list of service names.

muhamadazmy replied:
Why you still provide the service name? this should be in the filters.

muhamadazmy commented on pkg/zinit/commands.go:
```diff
@@ -304,3 +323,29 @@ func (c *Client) Forget(service string) error {
+
+	if err := c.StopWait(30*time.Second, service); err != nil {
+		return err
+	}
```
Because you can actually destroy multiple services (check the comments above) it means you need to use the StopAll method instead

Again, we are trying to make this as generic as possible and not crafted for a specific use case

abom committed:
move Stop/Start multiple methods from upgrader to zinit...

and implement Matches/Destory for multiple services

abom committed:
fix a typo in method docstring

abom committed:
update Matches/Destory to work with names only

abom committed:
update doc string of Destory()

abom committed:
use labels instead of allMatched flag

abom committed:
remove debug log

abom committed:
make sure service is migrated in networkd too...

create a reusable dhcp client service and use it in both
internet, networkd and dhcp monitor

also, make sure the exec check matches all parts (not only the first)

abom committed:
fix older dhcp client matching...

suggested by @muhamadazmy to add a regex matched for the service exec

abom committed:
dhcp monitor: destory older services at first

abom committed:
fix vet issues
